salt: Add kubelet_scrape_timeout to configure kubelet service monitor scrape timeout

JBWatenbergScality opened the pull request:
**Component**:

salt

**Context**: 

When using a bunch of sparseloopdevice kubelet metrics endpoint may spend
a lot of time to retrieve volumes metrics. In such situation the default
10s scrape timeout of Prometheus is not enough as the metrics endpoint may
take longer to answer.

To avoid losing data in such situations, we enabled customization of kubelet
scarpe timeout.

JBWatenbergScality committed:
salt: Add kubelet_scrape_timeout to configure kubelet service monitor configuration

bert-e commented:
# Hello jbwatenbergscality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @lucieleonard

* @teddyandrieux

* @thomasdanan

* @nicolast

* @chengyanjin

* @jbwatenbergscality

* @gdemonet

* @alexandre-allard-scality

alexandre-allard commented on charts/render.py:
```diff
@@ -254,6 +254,13 @@ def patch_grafana_dashboards(manifest):
     return manifest
 
 
+def patch_kubelet_service_monitor(manifest):
+    manifest["spec"]["endpoints"][0][
+        "scrapeTimeout"
+    ] = "__var__(prometheus.spec.config.kubelet_scrape_timeout)"
+    return manifest
```
You should rather change the `charts/kube-prometheus-stack.yaml` values:
```
kubelet:
  serviceMonitor:
    scrapeTimeout: "__var__(prometheus.spec.config.kubelet_scrape_timeout)"
```
And let the script handle that for you, without even touching it.

JBWatenbergScality replied:
👍  for some reason I hadn't found it configurable in the chart last week but in fact it is (https://github.com/scality/metalk8s/blob/development/2.11/charts/kube-prometheus-stack/templates/exporters/kubelet/servicemonitor.yaml#L45) so yes I can change it to what you suggested, good catch !!

JBWatenbergScality replied:
Done in 9ca67a0442b2eb1dee63f9b5bed84eda288c9e03

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @lucieleonard

* @teddyandrieux

* @thomasdanan

* @nicolast

* @chengyanjin

* @jbwatenbergscality

* @gdemonet

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @alexandre-allard-scality

alexandre-allard commented on docs/operation/cluster_and_service_configuration.rst:
```diff
@@ -324,10 +324,41 @@ This functionality can be actived:
 
 Then :ref:`apply the configuration<csc-prometheus-apply-cfg>`.
 
+Kubelet metrics scrape tiemout and both size and time based retentions can be
+activated at the same time.
```
For me this information is only relevant for the retention, because it could be mutually exclusive, you should let it where it was.

JBWatenbergScality replied:
Done in efb76c1cae80a2eda688435e521bea325ee75e98

alexandre-allard commented:
To apply the changes, the Prometheus manifest must be re-generated with the following command:
```
./charts/render.py prometheus-operator \
       charts/kube-prometheus-stack.yaml \
       charts/kube-prometheus-stack/ \
       --namespace metalk8s-monitoring \
       --service-config grafana \
       metalk8s-grafana-config \
       metalk8s/addons/prometheus-operator/config/grafana.yaml \
       metalk8s-monitoring \
       --service-config prometheus \
       metalk8s-prometheus-config \
       metalk8s/addons/prometheus-operator/config/prometheus.yaml \
       metalk8s-monitoring \
       --service-config alertmanager \
       metalk8s-alertmanager-config \
       metalk8s/addons/prometheus-operator/config/alertmanager.yaml \
       metalk8s-monitoring \
       --service-config dex \
       metalk8s-dex-config \
       metalk8s/addons/dex/config/dex.yaml.j2 metalk8s-auth \
       --drop-prometheus-rules charts/drop-prometheus-rules.yaml \
       > salt/metalk8s/addons/prometheus-operator/deployed/chart.sls
```

alexandre-allard commented on docs/operation/cluster_and_service_configuration.rst:
```diff
@@ -324,10 +324,41 @@ This functionality can be actived:
+Set Kubelet metrics scrape tiemout
+""""""""""""""""""""""""""""""""""
+
+When using a bunch of sparseloopdevice kubelet metrics endpoint may spend
+a lot of time to retrieve volumes metrics. In such situation the default
+10s scrape timeout of Prometheus is not enough as the metrics endpoint may
+take longer to answer.
+
+To avoid losing data in such situations, we enabled customization of kubelet
+scarpe timeout, to customize it use the following configuration:
```
```suggestion
In some cases (e.g. when using a lot of sparse loop devices), the kubelet metrics
endpoint can be very slow to answer and the Prometheus' default 10s scrape
timeout may not be sufficient. 

To avoid timeouts and thus losing metrics, you can customize the scrape timeout as follows:
```

JBWatenbergScality replied:
Done in efb76c1cae80a2eda688435e521bea325ee75e98

alexandre-allard commented on docs/operation/cluster_and_service_configuration.rst:
```diff
@@ -324,10 +324,41 @@ This functionality can be actived:
+       kind: PrometheusConfig
+       spec:
+         config:
+          kubelet_scrape_timeout: 30s
```
bad indentation:
```suggestion
           kubelet_scrape_timeout: 30s
```

alexandre-allard replied:
I'm wondering if we should rather add an extra level for service monitor, rather than putting it like this, e.g.:
```
spec:
  config:
    serviceMonitor:
      kubelet:
        scrapeTimeout: 30s
```
Or even:
```
spec:
  config:
    kubelet:
      scrapeTimeout: 30s
```

JBWatenbergScality replied:
Done in efb76c1cae80a2eda688435e521bea325ee75e98

JBWatenbergScality committed:
salt: Apply prometheus chart changes

Run ./charts/render.py prometheus-operator \
       charts/kube-prometheus-stack.yaml \
       charts/kube-prometheus-stack/ \
       --namespace metalk8s-monitoring \
       --service-config grafana \
       metalk8s-grafana-config \
       metalk8s/addons/prometheus-operator/config/grafana.yaml \
       metalk8s-monitoring \
       --service-config prometheus \
       metalk8s-prometheus-config \
       metalk8s/addons/prometheus-operator/config/prometheus.yaml \
       metalk8s-monitoring \
       --service-config alertmanager \
       metalk8s-alertmanager-config \
       metalk8s/addons/prometheus-operator/config/alertmanager.yaml \
       metalk8s-monitoring \
       --service-config dex \
       metalk8s-dex-config \
       metalk8s/addons/dex/config/dex.yaml.j2 metalk8s-auth \
       --drop-prometheus-rules charts/drop-prometheus-rules.yaml \
       > salt/metalk8s/addons/prometheus-operator/deployed/chart.sls

alexandre-allard reviewed (approved):
LGTM

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @lucieleonard

* @teddyandrieux

* @thomasdanan

* @nicolast

* @chengyanjin

* @jbwatenbergscality

* @gdemonet

* @alexandre-allard-scality

JBWatenbergScality commented:
/approved

bert-e commented:
# Unknown command



I didn't understand this comment by @jbwatenbergscality:

> /approved

I don't know what `approved` means.

Please **edit** or **delete** the corresponding comment so I can move on.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @lucieleonard

* @teddyandrieux

* @thomasdanan

* @nicolast

* @chengyanjin

* @jbwatenbergscality

* @gdemonet

* @alexandre-allard-scality

JBWatenbergScality commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.11`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.11`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye jbwatenbergscality.
